### DISCUSSION TITLE
Real time update of functions

### DISCUSSION THREAD

**Discussion by HenYu-Yo** [ManimCommunity/manim#3337] [Q&A] [answered]
```python
from manim import *

class demo(Scene):
    def construct(self):
        axe = Axes(
            x_range=(-5, 5),
            y_range=(-2, 2),
            x_length=10,
            y_length=5,
            axis_config={
                "include_numbers": True,
                "tip_shape": StealthTip
            }
        )
       w = ValueTracker(1)
       p = ValueTracker(0)
       def sin(x):
            return np.sin(w.get_value*x + p.get_value)
       graph = axe.plot(sin, x_range=(-5, 5), color=RED)
       self.add(axe, graph)
```
The above is the code. I originally intended to use the updater to achieve real-time changes in the function image with linear growth of two parameters. I know this can be achieved, but I cannot achieve it.

**Comment by uwezi** [comment, marked as answer]
Can you please come over to [Discord](https://docs.manim.community/en/stable/faq/general.html?highlight=discord#where-can-i-find-more-resources-for-learning-manim)? It's our preferred way to help people.

Also: what editor are you using for your code? Because you have a mismatch in indentation levels which should have been picked up easily by any modern editor for python code.
![image](https://github.com/ManimCommunity/manim/assets/8582807/e8b434d1-8254-4102-a476-d05098894106)

next error of yours: `.get_value()` is a function without arguments/parameters, yet it still needs a pair of parentheses.

Fixing these problems we get
```python
class demo(Scene):
    def construct(self):
        axe = Axes(
            x_range=(-5, 5),
            y_range=(-2, 2),
            x_length=10,
            y_length=5,
            axis_config={
                "include_numbers": True,
                "tip_shape": StealthTip
            }
        )
        w = ValueTracker(1)
        p = ValueTracker(0)
        def sin(x):
            return np.sin(w.get_value()*x + p.get_value())
        graph = axe.plot(sin, x_range=(-5, 5), color=RED)
        self.add(axe, graph) 
```
and an output like
![image](https://github.com/ManimCommunity/manim/assets/8582807/9f0cde5d-df38-48e3-9c38-46f6eb796851)

But in order to animate things you need to 

1. make sure that your plot is redrawn for every change in its parameters
2. change the values of your value trackers inside an animation

For (1) we will put the `.plot` into an `always_redraw()` and for the second use `self.play(w.animate.set_value())`

```python
class demo(Scene):
    def construct(self):
        axe = Axes(
            x_range=(-5, 5),
            y_range=(-2, 2),
            x_length=10,
            y_length=5,
            axis_config={
                "include_numbers": True,
                "tip_shape": StealthTip
            }
        )
        w = ValueTracker(1)
        p = ValueTracker(0)
        def sin(x):
            return np.sin(w.get_value()*x + p.get_value())
        graph = always_redraw(lambda:
            axe.plot(sin, x_range=(-5, 5), color=RED)
        )
        self.add(axe, graph) 
        self.wait()
        self.play(w.animate.set_value(2))
        self.wait()
        self.play(p.animate.set_value(-180*DEGREES))
        self.wait()
```

https://github.com/ManimCommunity/manim/assets/8582807/986a374d-e4ac-41d8-b9cb-f70f784b71e1

**Reply by HenYu-Yo** [reply]
抱歉，因为源文件有太多代码，我只挑取了重要部分，其中一部分是我在github上直接书写的。在实际过程中我并没有发生这些错误。
最后，感谢你的帮助
